File: apps/api/src/browser-session.ts
/** HttpOnly WorkOS browser-session cookie and CSRF boundary. */

import type http from "http";

import { signSignedToken, verifySignedToken } from "@janusly/engine/src/secrets";

import { isAllowedRequestOrigin } from "./http";

export const SSO_SESSION_PURPOSE = "sso_session" as const;
export const SESSION_COOKIE_NAME = "janusly_session";
export const SESSION_CSRF_HEADER = "x-janusly-csrf";

type BrowserSessionTokenPayload = { sessionId: string };

function parseCookies(header: string | string[] | undefined): Map<string, string> {
  const source = Array.isArray(header) ? header.join(";") : header ?? "";
  const cookies = new Map<string, string>();
  for (const part of source.split(";")) {
    const separator = part.indexOf("=");
    if (separator <= 0) continue;
    const name = part.slice(0, separator).trim();
    const value = part.slice(separator + 1).trim();
    if (!name || !value) continue;
    try { cookies.set(name, decodeURIComponent(value)); } catch { /* malformed cookie is ignored */ }
  }
  return cookies;
}

function secureCookie(): boolean {
  const explicit = process.env.JANUSLY_SESSION_COOKIE_SECURE;
  if (explicit === "true") return true;
  if (explicit === "false") return false;
  try {
    return new URL(process.env.JANUSLY_WEB_BASE_URL ?? "").protocol === "https:";
  } catch {
    return false;
  }
}

/** Mint a signed opaque browser token and its server-side expiry instant. */
export function createBrowserSessionToken(sessionId: string, ttlSeconds: number): {
  token: string;
  expiresAt: Date;
} {
  return {
    token: signSignedToken({
      purpose: SSO_SESSION_PURPOSE,
      payload: { sessionId },
      ttlSeconds,
    }),
    expiresAt: new Date(Date.now() + ttlSeconds * 1000),
  };
}

/** Verify and extract the opaque session id from the request cookie. */
export function readBrowserSessionId(req: http.IncomingMessage): string | null {
  const token = parseCookies(req.headers.cookie).get(SESSION_COOKIE_NAME);
  if (!token) return null;
  try {
    const envelope = verifySignedToken<typeof SSO_SESSION_PURPOSE, BrowserSessionTokenPayload>(
      token,
      SSO_SESSION_PURPOSE,
    );
    return typeof envelope.payload.sessionId === "string" && envelope.payload.sessionId.length > 0
      ? envelope.payload.sessionId
      : null;
  } catch {
    return null;
  }
}

/** Serialize the session cookie for a successful SSO callback or org switch. */
export function sessionCookie(token: string, ttlSeconds: number): string {
  return [
    `${SESSION_COOKIE_NAME}=${encodeURIComponent(token)}`,
    "Path=/",
    "HttpOnly",
    "SameSite=Lax",
    `Max-Age=${Math.max(0, Math.floor(ttlSeconds))}`,
    ...(secureCookie() ? ["Secure"] : []),
  ].join("; ");
}

/** Expire the browser cookie immediately. */
export function clearSessionCookie(): string {
  return [
    `${SESSION_COOKIE_NAME}=`,
    "Path=/",
    "HttpOnly",
    "SameSite=Lax",
    "Max-Age=0",
    ...(secureCookie() ? ["Secure"] : []),
  ].join("; ");
}

/**
 * Require Janusly's custom mutation header and an allowlisted browser origin.
 * This complements SameSite=Lax and applies only to cookie-authenticated writes.
 */
export function requireBrowserCsrf(req: http.IncomingMessage): void {
  const marker = req.headers[SESSION_CSRF_HEADER];
  const origin = req.headers.origin;
  if (marker !== "1" || typeof origin !== "string" || !isAllowedRequestOrigin(origin)) {
    const error = new Error("Forbidden: invalid browser session origin") as Error & { statusCode?: number };
    error.statusCode = 403;
    throw error;
  }
}

File: internal/browsersession/browsersession.go
// Package browsersession owns the opaque browser token, cookie attributes,
// and CSRF marker boundary used by cookie-authenticated routes.
package browsersession

import (
	"errors"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"time"

	"github.com/johnny4young/janusly/internal/signedtoken"
)

const (
	Purpose    = "sso_session"
	CookieName = "janusly_session"
	CSRFHeader = "x-janusly-csrf"
)

// ErrInvalidCSRF is intentionally uniform across missing marker and origin.
var ErrInvalidCSRF = errors.New("Forbidden: invalid browser session origin") //nolint:staticcheck // contract message is the wire contract

type tokenPayload struct {
	SessionID string `json:"sessionId"`
}

// Token is the signed opaque value plus its exact signed expiry.
type Token struct {
	Value     string
	ExpiresAt time.Time
}

// CreateToken signs only the server-side session row id.
func CreateToken(sessionID string, ttlSeconds int) (Token, error) {
	if sessionID == "" {
		return Token{}, signedtoken.ErrInvalid
	}
	value, envelope, err := signedtoken.Sign(Purpose, tokenPayload{SessionID: sessionID}, ttlSeconds)
	if err != nil {
		return Token{}, err
	}
	return Token{Value: value, ExpiresAt: time.Unix(envelope.ExpiresAt, 0)}, nil
}

// ReadSessionID authenticates and extracts the opaque id from request cookies.
func ReadSessionID(r *http.Request) string {
	cookie, err := r.Cookie(CookieName)
	if err != nil || cookie.Value == "" {
		return ""
	}
	envelope, err := signedtoken.Verify[tokenPayload](cookie.Value, Purpose)
	if err != nil || envelope.Payload.SessionID == "" {
		return ""
	}
	return envelope.Payload.SessionID
}

// SessionCookie serializes the successful-login or organization-switch cookie.
func SessionCookie(token string, ttlSeconds int) string {
	return cookieValue(token, max(ttlSeconds, 0))
}

// ClearCookie expires the browser cookie immediately.
func ClearCookie() string {
	return cookieValue("", 0)
}

func cookieValue(token string, ttlSeconds int) string {
	value := CookieName + "=" + url.QueryEscape(token) +
		"; Path=/; HttpOnly; SameSite=Lax; Max-Age=" + strconv.Itoa(ttlSeconds)
	if secureCookie() {
		value += "; Secure"
	}
	return value
}

// SecureCookie exposes the deployment's cookie posture so sibling flows
// (the SSO state binding) inherit it instead of re-deriving it.
func SecureCookie() bool { return secureCookie() }

func secureCookie() bool {
	switch os.Getenv("JANUSLY_SESSION_COOKIE_SECURE") {
	case "true":
		return true
	case "false":
		return false
	}
	base, err := url.Parse(os.Getenv("JANUSLY_WEB_BASE_URL"))
	return err == nil && base.Scheme == "https"
}

// RequireCSRF requires both Janusly's custom marker and a concrete allowed
// origin. The caller supplies the same origin policy used by CORS.
func RequireCSRF(r *http.Request, originAllowed func(string) bool) error {
	if originAllowed == nil || r.Header.Get(CSRFHeader) != "1" ||
		!originAllowed(r.Header.Get("Origin")) {
		return ErrInvalidCSRF
	}
	return nil
}
